### PR TITLE
Fix AutoencoderTiny encoder scaling convention

### LINKED ISSUE #4676: TAESD-encoded latents are too dark
### Describe the bug

AutoencodeTiny (TAESD) _decoder_ seems to work fine. _encoding_ on the other hand is producing poor results, and an encode-decode round-trip turns out poorly:

input:
benz

output:
sad benz

### Reproduction

see https://gist.github.com/keturn/b0a10a3b388e1e49cdf38567b76eb30c

```py
import diffusers, torch
from PIL.Image import Image, open as image_open

device = torch.device("cuda:0")

with torch.inference_mode():
    taesd = diffusers.AutoencoderTiny.from_pretrained("madebyollin/taesd", torch_dtype=torch.float16).to(device=device)
    vaesd = diffusers.AutoencoderKL.from_pretrained("runwayml/stable-diffusion-v1-5", subfolder="vae", variant="fp16", torch_dtype=torch.float16).to(device=device)

from diffusers.utils.testing_utils import load_image

image = load_image(
    "https://huggingface.co/datasets/hf-internal-testing/diffusers-images/resolve/main/versatile_diffusion/benz.jpg"
)

from diffusers.image_processor import VaeImageProcessor
vae_processor = VaeImageProcessor()
image_tensor: torch.FloatTensor = vae_processor.preprocess(image).to(dtype=torch.float16, device=device)

print(f"image tensor range: {image_tensor.min()} < {image_tensor.mean()} < {image_tensor.max()})")


with torch.inference_mode():
    taesd_latents = taesd.encode(image_tensor).latents
    print(f"taesd-encoded latent range: {taesd_latents.min()} < {taesd_latents.mean()} (σ={taesd_latents.std()}) < {taesd_latents.max()})")

    vaesd_latents = vaesd.encode(image_tensor).latent_dist.sample()
    print(f"vaesd-encoded latent range: {vaesd_latents.min()} < {vaesd_latents.mean()} (σ={vaesd_latents.std()}) < {vaesd_latents.max()})")


with torch.inference_mode():
    redecoded_tensor = taesd.decode(taesd_latents).sample


redecoded_image = vae_processor.postprocess(redecoded_tensor)
display(image, redecoded_image[0])


from diffusers.commands import env
env.EnvironmentCommand().run()
```

### System Info

  - `diffusers` version: 0.20.0
  - Platform: Linux-5.15.0-79-generic-x86_64-with-glibc2.35
  - Python version: 3.11.4
  - PyTorch version (GPU?): 2.0.1+cu118 (True)
  - Huggingface_hub version: 0.16.4
  - Transformers version: 4.31.0
  - Accelerate version: 0.21.0
  - xFormers version: 0.0.21
  - Using GPU in script?: yes
  - Using distributed or parallel set-up in script?: no

### Who can help?

_No response_

### PR DESCRIPTION
**Pull request by madebyollin**
# What does this PR do?

  * Add [-1, 1] -> [0, 1] rescaling to `EncoderTiny`

  * Move [0, 1] -> [-1, 1] rescaling from `AutoencoderTiny.decode` to `DecoderTiny` (i.e. immediately after the final conv, as early as possible)

  * Fix missing [0, 255] -> [0, 1] rescaling in `AutoencoderTiny.forward`

  * Update `AutoencoderTinyIntegrationTests` to protect against scaling issues. The new test constructs a simple image, round-trips it through `AutoencoderTiny`, and confirms the decoded result is approximately equal to the source image. This test checks behavior with and without tiling enabled. This test will fail if new `AutoencoderTiny` scaling issues are introduced.

## Motivation

Raw TAESD weights expect images in [0, 1], but diffusers' convention represents images with zero-centered values in [-1, 1], so `AutoencoderTiny` needs to scale / unscale images at the start of encoding and at the end of decoding in order to work with diffusers.

Fixes #4676

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [ ] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: tests/models/test_models_vae.py [hunk 1]
```diff
@@ -302,19 +294,27 @@ def test_tae_tiling(self, in_shape, out_shape):
...
+            model.enable_tiling()
+
+        # make a black image with a white square in the middle,
+        # which is large enough to split across multiple tiles
+        image = -torch.ones(1, 3, 1024, 1024, device=torch_device)
+        image[..., 256:768, 256:768] = 1.0
 
+        # round-trip the image through the autoencoder
         with torch.no_grad():
             sample = model(image).sample
```
Comment (keturn) [inline tests/models/test_models_vae.py:312]: The test is much appreciated! But this isn't a round-trip, is it?

Comment (madebyollin) [reply tests/models/test_models_vae.py:312]: `AutoencoderTiny.forward` runs encode -> quantize -> dequantize -> decode internally https://github.com/madebyollin/diffusers/blob/main/src/diffusers/models/autoencoder_tiny.py#L321, so this test exercises both encoding and decoding

Comment (keturn) [reply tests/models/test_models_vae.py:312]: It does what now? Oh, so it does, I see. Is that how one trains an autoencoder or something? All good then.

File: tests/models/test_models_vae.py [hunk 2]
```diff
@@ -270,14 +270,6 @@ def tearDown(self):
         gc.collect()
         torch.cuda.empty_cache()
 
-    def get_file_format(self, seed, shape):
-        return f"gaussian_noise_s={seed}_shape={'_'.join([str(s) for s in shape])}.npy"
-
-    def get_sd_image(self, seed=0, shape=(4, 3, 512, 512), fp16=False):
-        dtype = torch.float16 if fp16 else torch.float32
-        image = torch.from_numpy(load_hf_numpy(self.get_file_format(seed, shape))).to(torch_device).to(dtype)
-        return image
```
Comment (sayakpaul) [inline tests/models/test_models_vae.py:279]: Could I get an explanation on why is this going away?

File: tests/models/test_models_vae.py [hunk 3]
```diff
@@ -302,19 +294,27 @@ def test_tae_tiling(self, in_shape, out_shape):
...
-        assert sample.shape == image.shape
-
-        output_slice = sample[-1, -2:, -2:, :2].flatten().float().cpu()
-        expected_output_slice = torch.tensor([0.9858, 0.9262, 0.8629, 1.0974, -0.091, -0.2485, 0.0936, 0.0604])
+        # the autoencoder reconstruction should match original image, sorta
+        def downscale(x):
+            return torch.nn.functional.avg_pool2d(x, model.spatial_scale_factor)
 
-        assert torch_all_close(output_slice, expected_output_slice, atol=3e-3)
+        assert torch_all_close(downscale(sample), downscale(image), atol=0.125)
```
Comment (sayakpaul) [inline tests/models/test_models_vae.py:317]: We prefer to keep our tests simpler than this. So, I would very much appreciate if we could keep the existing tests as untouched as possible and instead, add a new test for the round-trip behaviour.

**Comment by sayakpaul** [comment]
The failing tests seem to be irrelevant.

File: tests/models/test_models_vae.py (hunk 3 above)
Comment (madebyollin) [reply tests/models/test_models_vae.py]: Makes sense - I'll keep the existing test too (and just update the golden values).

File: tests/models/test_models_vae.py (hunk 2 above)
Comment (madebyollin) [reply tests/models/test_models_vae.py]: Thanks for reviewing! I had removed these because they weren't needed for the new test - but re-added them now.

File: tests/models/test_models_vae.py [hunk 4]
```diff
@@ -312,10 +312,32 @@ def test_stable_diffusion(self):
         assert sample.shape == image.shape
 
         output_slice = sample[-1, -2:, -2:, :2].flatten().float().cpu()
-        expected_output_slice = torch.tensor([0.9858, 0.9262, 0.8629, 1.0974, -0.091, -0.2485, 0.0936, 0.0604])
+        expected_output_slice = torch.tensor([0.0093, 0.6385, -0.1274, 0.1631, -0.1762, 0.5232, -0.3108, -0.0382])
```
Comment (sayakpaul) [inline tests/models/test_models_vae.py:315]: Sorry, I don't follow. Why are these values being changed? The decoded output shouldn't change no? What am I missing out on?

File: src/diffusers/models/autoencoder_tiny.py [hunk 5]
```diff
@@ -334,7 +331,7 @@ def forward(
         """
         enc = self.encode(sample).latents
         scaled_enc = self.scale_latents(enc).mul_(255).round_().byte()
-        unscaled_enc = self.unscale_latents(scaled_enc)
+        unscaled_enc = self.unscale_latents(scaled_enc / 255.0)
```
Comment (sayakpaul) [inline src/diffusers/models/autoencoder_tiny.py:334]: Worth adding a comment on why we're scaling here since our pixel normalization / de-normalization logic is handled in the `image_processor.py` module.

Comment (madebyollin) [reply src/diffusers/models/autoencoder_tiny.py:334]: The `/ 255.0` here is necessary to invert the `.mul_(255).round_().byte()` on the preceding line. In https://github.com/madebyollin/taesd/blob/main/taesd.py#L86, the division by 255 is [handled](https://github.com/pytorch/vision/blob/main/torchvision/transforms/functional.py#L174) automatically by torchvision's `to_tensor`, but this code doesn't use `to_tensor` so we need to apply the division ourselves.

Comment (sayakpaul) [reply src/diffusers/models/autoencoder_tiny.py:334]: Ah cool. Maybe worth adding a comment explaining it :)

File: tests/models/test_models_vae.py (hunk 4 above)
Comment (madebyollin) [reply tests/models/test_models_vae.py:315]: Indeed - the decoder behavior already worked fine (and this PR doesn't change it). However, this test doesn't just run the decoder; this test does a full round-trip through `AutoencoderTiny.forward` (which includes the encoder, as well as quantization / dequantization of latents).

The existing round trip code had two bugs (fixed in this PR):
1. No  [-1, 1] -> [0, 1] rescaling of the inputs in EncoderTiny
2. No [0, 255] -> [0, 1] rescaling in AutoencoderTiny.forward

It looks like the golden values in this test were computed using the previous buggy round-trip code, so because I've fixed the bugs in the round-trip code, I also need to change the golden values in this test.

File: src/diffusers/models/autoencoder_tiny.py (hunk 5 above)
Comment (madebyollin) [reply src/diffusers/models/autoencoder_tiny.py:340]: Done, hopefully it's clearer now

File: tests/models/test_models_vae.py (hunk 4 above)
Comment (sayakpaul) [reply tests/models/test_models_vae.py:315]: Thanks for explaining. Now it all makes sense to me! I will get the golden values for you from the internal CI runner to ensure exact matches.

Comment (sayakpaul) [reply tests/models/test_models_vae.py:315]: Ah it's passing already, great then!

**Review by sayakpaul — APPROVED** [review]
Thanks so much! 

@yiyixuxu @DN6 could one of you give this a look as well?

**Review by yiyixuxu — APPROVED** [review]
thank you!
